src/App.js

```diff
@@ -29,37 +29,115 @@ const App = () => {
   // This starts state off as a 2D array of JS objects with
   // empty value and unique ids.
   const [squares, setSquares] = useState(generateSquares());
+  const [winner, setWinner] = useState(null);
+  const [currentPlayer, setCurrentPlayer] = useState(PLAYER_1);
 
   // Wave 2
   // You will need to create a method to change the square
   //   When it is clicked on.
   //   Then pass it into the squares as a callback
 
   const checkForWinner = () => {
-    // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if
-    //    all three squares have the same value.
+    let i = 0;
+
+    // Check all the rows and columns for a winner
+    while (i < 3) {
+      if (
+        squares[i][0].value === squares[i][1].value &&
+        squares[i][2].value === squares[i][1].value &&
+        squares[i][0].value !== ''
+      ) {
+        return squares[i][0].value;
+      } else if (
+        squares[0][i].value === squares[1][i].value &&
+        squares[2][i].value === squares[1][i].value &&
+        squares[0][i].value !== ''
+      ) {
+        return squares[0][i].value;
+      }
+      i += 1;
+    }
+    // Check Top-Left to bottom-right diagonal
+    if (
+      squares[0][0].value === squares[1][1].value &&
+      squares[2][2].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][0].value;
+    }
+
+    // Check Top-right to bottom-left diagonal
+    if (
+      squares[0][2].value === squares[1][1].value &&
+      squares[2][0].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][2].value;
+    }
+
+    return null;
+  };
+
+  // const onClickCallBack = (id) => {
+  //   const emptyBoard = [...squares];
+  //   for (let row of emptyBoard) {
+  //     for (let square of row) { 
+  //       if (square.id === id && square.value === '') {
+  //         square.value = currentPlayer;
+  //         if (winner) {
+  //           !setCurrentPlayer();
+  //           checkForWinner();
+  //           setSquares(emptyBoard);
+  //         } else if (!winner) {
+  //           if (currentPlayer === PLAYER_1) {
+  //             setCurrentPlayer(PLAYER_2);
+  //           } else if (currentPlayer === PLAYER_2) {
+  //             setCurrentPlayer(PLAYER_1);
+  //           }
+  //         }
+  //       }
+  //     }
+  //   }
+  // };
+
+  const onClickCallback = (id) => {
+    setSquares((squares) => {
```

Comment:
Yes, we can use use the callback style of updating the squares here, but we don't actually know whether we want to or not. What if the player clicked on a square that was already played on? What if the game was already over? In that case, we could track whether a play was made, and if not, simply return the original board reference.